### ISSUE TITLE
Images & formatting lost after writing to existing file

### ISSUE THREAD

**Issue by edithemmings** [zdavatz/spreadsheet#275]
I have the following two models in a Rails app:
```
class ModelA
  has_one_attached :template_file
end

class ModelB
  has_one_attached :export_file
end
```
My goal is to make a copy of the template file from ModelA, write data to that copy, and save it to the export file on ModelB. 
My ModelA template file is an xls attachment with 4 worksheets. The first worksheet has a styled header row. I am writing only to the columns under those headers. The other worksheets in the file have images in them. 

The result of my code is a copy of the file with the same worksheets with my data written to the first sheet as-expected--except all images are missing and some of the header formatting is lost (for example, text alignment is lost, but background color is not).

If I comment out the call to `write_to_sheet(sheet)` in the `build_and_attach_export` method, then I end up with a perfect copy of the original file saved as `export_file`.  However any attempt to write to any cell in the sheet results in the lost images and formatting. 

```ruby
def build_and_attach_export
  # pull down template/copy to tempfile (this is working fine)
  template = model_a.template_file
  filename_base = template.blob.filename.base
  filename_ext = template.blob.filename.extension_with_delimiter
  tmp = Tempfile.new([filename_base, filename_ext], encoding: 'ascii-8bit')
  tmp.write(template.download)
  tmp.rewind

  # parse tmp file using Spreadsheet (also working fine)
  book = Spreadsheet.open(tmp)
  tmp.unlink
  sheet = book.worksheet(0)

  # This seems to be the origin of my problems. Yes, it writes to the file, but it messes up other elements of the file.
  write_to_sheet(sheet)

  # write to IO for upload (also working fine)
  io = StringIO.new
  book.write(io)
  io.rewind
  blob = ActiveStorage::Blob.create_and_upload!(io: ,  filename: 'export.xls',  content_type: 'application/vnd.ms-excel')
  model_b.export_file.attach(blob)
end

def write_to_sheet(sheet)
  # note: source_data is an array of arrays(rows) of strings(cell values)
  source_data.each_with_index do |row_vals, row_idx|
    row_vals.each_with_index do |val, col_idx|
      next unless val

      # I tried some different syntax from the Spreadsheet docs, but any attempt to write to the file messes up my images/formatting
      sheet.rows[row_idx + 1][col_idx] = val
    end
  end
end
```

Does anyone have any insight into why this might be happening?

**Comment by zdavatz** [comment]
Have you tried writing to a new file? Does that work?

**Comment by edithemmings** [comment]
> Have you tried writing to a new file? Does that work?

@zdavatz Could you be more specific? It looks like you're suggesting writing my data to a new file instead of the tmp copy of the template, which would defeat the purpose of what I am trying to do. It is the existing images and styling in the template that are getting lost when I write to unrelated cells/sheets.

**Comment by edithemmings** [comment]
> which software created the original file you are trying to edit with the Spreadsheet Gem?

The original template file was made manually by a human. The file that I am parsing in Spreadsheet and writing to is a Tempfile that I wrote the original file into (variable `tmp` in the above example).

**Comment by zdavatz** [comment]
Was the Original File created by Openoffice or MS Office? On Mac or on Windows? Any chances of seeing the original file?

**Comment by edithemmings** [comment]
> Was the Original File created by Openoffice or MS Office? On Mac or on Windows? Any chances of seeing the original file?

I'm not able to share the original file unfortunately. And I do not know the circumstances of its original creation. My client has many template files like this that have existed for a long time. If there are ways they could re-create the templates for a better chance of success, then I am all ears.

**Comment by zdavatz** [comment]
I would try to created the templates with Ruby spreadsheet from the start, then modify the file and then save as a new file. Sometimes the software creating the original files adds binary information, that Ruby Spreadsheet does not add, then all fails. So if you can, I would go the whole way with Ruby Spreadsheet.

**Comment by edithemmings** [comment]
> I would try to created the templates with Ruby spreadsheet from the start, then modify the file and then save as a new file. Sometimes the software creating the original files adds binary information, that Ruby Spreadsheet does not add, then all fails. So if you can, I would go the whole way with Ruby Spreadsheet.

Ok that could explain it, but I'm still not understanding why this works perfectly:
```
book = Spreadsheet.open(tmp)
io = StringIO.new
book.write(io)
io.rewind
```
But this destroys the binary info:
```
book = Spreadsheet.open(tmp)
sheet = book.worksheet(0)
sheet.rows[1][0] = 'hello'
io = StringIO.new
book.write(io)
io.rewind
```
What am I missing?

**Comment by edithemmings** [comment]
> Can we try to reproduce this with a sample file of yours?

I was able to reproduce with this template file:
[template.xls](https://github.com/zdavatz/spreadsheet/files/15179725/template.xls)

And here is the resulting export file with "hello" written to it. The image on the second sheet has been lost. 
[export.xls](https://github.com/zdavatz/spreadsheet/files/15179743/export.xls)

**Comment by zdavatz** [comment]
Ok, so the original file was created by Miguel Rocafort. Which software did he use to create the file? Which OS? Did he use the spreadsheet gem to the create the file template.xls?

**Comment by edithemmings** [comment]
> Ok, so the original file was created by Miguel Rocafort. Which software did he use to create the file? Which OS? Did he use the spreadsheet gem to the create the file template.xls?

Not sure who that is, but the spreadsheet gem was not used to create the template file. These templates are created by users -- not with this gem, so using the gem to produce test templates does not match my use case. I used this chrome extension to create the above test file: https://chromewebstore.google.com/detail/xls-editor/iobjaooppmgjlgomfpaohhncpfjpigaf

From what I understood you were looking to see the behavior reproduced, and the above test file matches what I'm seeing with my users' files.

**Comment by edithemmings** [comment]
It’s looking like this gem may not be the right fit for my case. I wish I had a deeper understanding (and more time) so I could contribute to researching, but realistically at this point I just need a workaround. If I can get my client on board with converting template files to xlsx, then I can use the RubyXL gem with the same logic, and the file remains intact.

Thanks so much for your time and effort! I’m happy to keep discussing/providing context if its of any further value.

**Comment by gregogalante** [comment]
Im getting the same problem. I take a .xls template generated by human, update some specific cells, save as new file. The new file has lost an image inside the template.

Thats my code:
```ruby
xls_file_path = Rails.root.join('vendor', 'rai', Rails.env, 'report_dotazione_estintori_m.xls')
    book = Spreadsheet.open(xls_file_path)
    sheet = book.worksheet(0)
    sheet[0, 3] = 'custom content'
    output_file_name = "#{output}.xls"
    output_file = Rails.root.join('tmp', output_file_name)
    book.write(output_file)
````

I don't know the program used to create the file but i edit it with MS Excel for Mac. I can share it if necessary

**Comment by zdavatz** [comment]
Yes, please share the file here together with your script.

**Comment by gregogalante** [comment]
Can i share the file privately? I can send you via email

**Comment by gregogalante** [comment]
@zdavatz This is my code (with exactly cells that i modify):

```ruby
def example_method
    xls_file_path = Rails.root.join('vendor', 'rai', Rails.env, 'report_dotazione_estintori_m.xls')

    book = Spreadsheet.open(xls_file_path)
    sheet = book.worksheet(0)
    # set client.code on D1
    sheet.rows[0][3] = 'code'
    # set client.final_name on D2
    sheet.rows[1][3] = 'final_name'
    # set client.address on D3
    sheet.rows[2][3] = 'address'
    # set date on OPQRS2
    sheet.rows[1][14] = Date.today.strftime('%d/%m/%Y')
    # set user.full_name on OPQRSTUVWXYZAA4/5
    sheet.rows[3][14] = 'full_name'

    # set maintenances data starting from row 12
    row = 11
    [1].each do |maintenance|
      sheet.rows[row][0] = 'xyz'
      sheet.rows[row][1] = 'xyz'
      sheet.rows[row][3] = 'xyz'
      sheet.rows[row][9] = 'xyz'
      sheet.rows[row][10] = 'xyz'
      sheet.rows[row][18] = 'xyz'
      sheet.rows[row][21] = 'xyz'
      sheet.rows[row][25] = 'xyz'
      sheet.rows[row][28] = 'xyz'
      sheet.rows[row][31] = 'xyz'
      sheet.rows[row][36] = 'xyz'
      sheet.rows[row][37] = 'xyz'
      sheet.rows[row][38] = 'xyz'
      sheet.rows[row][39] = 'xyz'

      row += 1
    end

    output_file_name = "#{Time.now.to_i}.xls"
    output_file = Rails.root.join('tmp', output_file_name)
    book.write(output_file)
  end
```

I send the xls on your email!

**Comment by zdavatz** [comment]
1. How was the original file created? Using which software? What OS?
2. ![Image](https://github.com/user-attachments/assets/9232b0f2-df27-4201-8a48-63cb57a883ed)
3. Did you try to first save the original file created in 2001, check the contents then edit it again and then save it?
4. A lot of things happend between 2001 and 2025 and there may have been changes to the Binary format.

**Comment by zdavatz** [comment]
How was the original file that you copy over created? Is that the file from 2001? Does it work, with a file that you do not copy over from a Template?

**Comment by gregogalante** [comment]
I have tried with a new file generated from scratch ([attached here](https://storage.gregoriogalante.com/lorem-ipsum/example.xls)) but the problem persists.

**Comment by zdavatz** [comment]
Ok, so you can not create a file from scratch, that contains images, using the gem, is that correct?

**Comment by gregogalante** [comment]
Yes! I need to open an existing xls that contains an image, change some cells values, and than save it on a new file.

**Comment by zdavatz** [comment]
That works?

**Comment by gregogalante** [comment]
No, it loose the image

**Comment by zdavatz** [comment]
So there is no way you can edit the spreadsheet, or create a new one, adding an image?

**Comment by gregogalante** [comment]
No, however i have tried to do the same thing with rubyXL on a xlsx and ti works!

**Comment by zdavatz** [comment]
Yes, xlsx is XML, xls is binary.
